03-Binary-Search-Trees/02-Binary-Search-Trees.md

```diff
@@ -191,480 +86,938 @@ class Tree:
 
-### Insertion - O(log n)
-
-The _root_ is where the tree begins, the topmost node.  New nodes as they are added are placed to the left or a given node, if they are less than or equal to the current node, and to the right if they are greater than the current node.  This is a natually recursive process.
-
-```
-Method add:
-  if the root is nil set the root to be a new node with the given value and return the node.
-
-  Otherwise:
-    if the value is less than or equal to the current node's value, make the current node's left be the result of calling add on root's left.
-    otherwise make node's right be the result of calling add on node's right.
-```
-
-![Tree Insert operation visualization](./images/Binary-search-trees__insert-into-tree.gif)
+### Searching
 
-_Fig.  Visualization of inserting a value into a BST_
+When we search for a node in a linked list, we start at our traversal at the head of the list. In a binary search tree, the equivalent of the head of the list is the root of the tree. The root is where the tree begins; the topmost node. Thus, when we search for a node in a binary search tree, we will always start our traversal at the root of the tree.
 
-You can experiment with this in the [Binary Tree Visualizer](https://visualgo.net/en/bst)
+Our search method will find the first node in the tree with a given key. Unlike a linked list, we won't have to search the entire tree to find the node we are looking for. Because the left subtree of a given node will only hold nodes with keys less than the given node, and the right subtree will only hold nodes with keys greater than or equal to the given node, we have to search through at most one of the root node's subtrees to find our node.
 
-### Deletion - O(log n)
+We can then look at the root of the relevant subtree and make the same decision about which of its own subtrees we should next traverse. 
 
-To delete a node from a binary search tree we must first find the node to delete.  This is an O(log n) operation.  Once we find the node we can delete it by changing the references.
+Notice that this is a naturally recursive process: We can call search on each subtree until we find the node we are searching for or reach a leaf in the tree. If we reach a leaf (our base case - a tree with a single node) and still haven't found the node we are searching for, we return None. 
 
-```
-Method delete(current_root, key):
-    if the root is none return none
-  
-    Otherwise, recur down the tree
-        if the key is less than the root's key delete the node in the left subtree.
-            current_root.left = delete(current_root.left, key);
-        otherwise if the key is greater than root's key delete the node in the right subtree.
-            current_root.right = delete(current_root.right, key);
+### Finding A Node With Python
 
+![Finding a 29 in a tree visualization](./images/Binary-Search-Trees__find-value.gif)
```

Comment:
I could certainly be mistaken here, but from what I can tell it seems that this GIF demonstrates adding a new node to a BST, rather than searching for a value.